### PR TITLE
Adds parser for external file downloads

### PR DESCRIPTION
**Pull request by elenatanasoiu**
Adds a parser for reading logs generated by the CDN. These logs contain information about which files are being downloaded from gov.uk via external links.

This information will be parsed and sent along to google analytics.

Tests have been written using nose.

### CODE REVIEW

File: download-lambda-logs/download_logs/aws_lambda.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        client_id = client_id.replace('GA', '')
+        client_id = client_id.split(".")[-2:]
+        return ".".join(client_id)
+
+    def send_event_to_GA(self, download_data):
+        PROPERTY_ID = 'UA-26179049-7'
+
+        params = urllib.parse.urlencode({
+                                        'v': 1,
+                                        'tid': PROPERTY_ID,
```
Comment (boffbowsh) [inline download-lambda-logs/download_logs/aws_lambda.py]: As this is just a local variable, it should be downcased. Spelling this like a constant makes me want to look elsewhere in the file or in config for it

File: download-lambda-logs/download_logs/tests/2017-05-26T10-00-00.000-wZe41G6PdJYziQ8AAAAA.log [hunk 2]
```diff
@@ -0,0 +1,3 @@
+1495792799	206	/government/uploads/system/uploads/attachment_data/file/224634/Children_travelling_to_the_UK_leaflet_A5_WEB_final.pdf	81.111.204.199		Mozilla/5.0 (Windows Phone 8.1; ARM; Trident/7.0; Touch; rv:11.0; IEMobile/11.0; Microsoft; Lumia 640 LTE) like Gecko	GA1.3.1721678609.1462462900
```
Comment (boffbowsh) [inline download-lambda-logs/download_logs/tests/2017-05-26T10-00-00.000-wZe41G6PdJYziQ8AAAAA.log]: Can you anonymise the IPs and GA Client IDs in this file please?

File: download-lambda-logs/download_logs/aws_lambda.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        self.bucket = bucket
+        super().__init__(filename)
+
+    def open_for_read(self):
+        print(self.bucket, self.filename)
+        obj = s3.get_object(Bucket=self.bucket, Key=self.filename)
+        return TextIOWrapper(gzip.GzipFile(fileobj=obj['Body'], mode='r'))
+
+    def process(self):
+        path = os.path.dirname(self.filename)
```
Comment (boffbowsh) [inline download-lambda-logs/download_logs/aws_lambda.py]: These aren't needed as you're not writing any new files

File: download-lambda-logs/download_logs/aws_lambda.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        y, m, d = date.split('-')
+
+        rest = rest.replace('.log.gz', '.json.gz')
+
+        csvreader = csv.reader(self.open_for_read(), delimiter="\t")
+
+        data = []
+        for row in csvreader:
+            result = self.transform_row(row)
+            if result != '' and result != None:
```
Comment (boffbowsh) [inline download-lambda-logs/download_logs/aws_lambda.py]: `result is not None`

File: download-lambda-logs/download_logs/aws_lambda.py (hunk 3 above)
Comment (elenatanasoiu) [inline download-lambda-logs/download_logs/aws_lambda.py]: Removed

File: download-lambda-logs/download_logs/aws_lambda.py (hunk 4 above)
Comment (elenatanasoiu) [inline download-lambda-logs/download_logs/aws_lambda.py]: Changed via flake8

File: download-lambda-logs/download_logs/aws_lambda.py (hunk 1 above)
Comment (elenatanasoiu) [inline download-lambda-logs/download_logs/aws_lambda.py]: downcased

File: download-lambda-logs/download_logs/aws_lambda.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        PROPERTY_ID = 'UA-26179049-7'
+
+        params = urllib.parse.urlencode({
+                                        'v': 1,
+                                        'tid': PROPERTY_ID,
+                                        'cid': download_data['ga_client_id'],
+                                        't' : 'event',
+                                        'ec': 'Document downloaded from external source',
+                                        'ea': download_data['file_downloaded'] or 'No filename present',
+                                        'el': download_data['referer'] or 'No referer',
```
Comment (elenatanasoiu) [inline download-lambda-logs/download_logs/aws_lambda.py]: Removed typo from everywhere

File: download-lambda-logs/download_logs/tests/2017-05-26T10-00-00.000-wZe41G6PdJYziQ8AAAAA.log (hunk 2 above)
Comment (elenatanasoiu) [inline download-lambda-logs/download_logs/tests/2017-05-26T10-00-00.000-wZe41G6PdJYziQ8AAAAA.log]: Anonymised

File: download-lambda-logs/download_logs/aws_lambda.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+s3 = boto3.client('s3')
+
+
+class AWSLambda(Base):
+    def __init__(self, bucket, filename):
+        self.bucket = bucket
+        super().__init__(filename)
+
+    def open_for_read(self):
+        print(self.bucket, self.filename)
```
Comment (boffbowsh) [inline download-lambda-logs/download_logs/aws_lambda.py]: Is this a debug statement?

File: download-lambda-logs/download_logs/aws_lambda.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        obj = s3.get_object(Bucket=self.bucket, Key=self.filename)
+        return TextIOWrapper(gzip.GzipFile(fileobj=obj['Body'], mode='r'))
+
+    def process(self):
+        name = os.path.basename(self.filename)
```
Comment (boffbowsh) [inline download-lambda-logs/download_logs/aws_lambda.py]: None of this is needed as far as I can see.

File: download-lambda-logs/download_logs/aws_lambda.py (hunk 6 above)
Comment (elenatanasoiu) [inline download-lambda-logs/download_logs/aws_lambda.py]: Plucked from here: https://github.com/alphagov/govuk-lambda-app-deployment/blob/lambda-logs/lambda-logs/govuk_logs/aws_lambda.py#L19

Removed now

File: download-lambda-logs/download_logs/aws_lambda.py (hunk 7 above)
Comment (elenatanasoiu) [inline download-lambda-logs/download_logs/aws_lambda.py]: Removed

**Pull request merged**